src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php

```diff
@@ -233,7 +233,13 @@ private function autowireMethod(\ReflectionFunctionAbstract $reflectionMethod, a
                     if ($parameter->isOptional()) {
                         continue;
                     }
-                    throw new AutowiringFailedException($this->currentId, sprintf('Cannot autowire service "%s": argument "$%s" of method "%s()" must have a type-hint or be given a value explicitly.', $this->currentId, $parameter->name, $class !== $this->currentId ? $class.'::'.$method : $method));
+                    if ($type = ProxyHelper::getTypeHint($reflectionMethod, $parameter, false)) {
+                        $message = 'Cannot autowire service "%s": argument "%s $%s" of method "%s()" must be given a value explicitly.';
+                    } else {
+                        $message = 'Cannot autowire service "%s": argument "%s$%s" of method "%s()" must have a type-hint or be given a value explicitly.';
```

Comment:
just adding a type hint does not neccessarily help like `string`. So maybe it should say something like "must have a known type-hint".